examples/seq2seq/distillation.py

```diff
@@ -156,26 +189,28 @@ def _step(self, batch):
         def zero_tensor():
             return torch.tensor(0.0).type_as(student_lm_loss)
 
+        teacher_enc_outputs = enc_outputs
         hid_loss_enc, hid_loss_dec = zero_tensor(), zero_tensor()
         if self.different_encoder:  # compute encoder hidden state loss
             with torch.no_grad():
-                teacher_enc_hid = self.teacher.get_encoder()(
-                    input_ids, attention_mask=src_mask, output_hidden_states=True, return_dict=True
-                ).hidden_states
-
-            hid_loss_enc = self.calc_hidden_loss(
-                src_mask,
-                enc_hidden_state,
-                teacher_enc_hid,
-                self.e_matches,
-                normalize_hidden=self.hparams.normalize_hidden,
-            )
-
+                teacher_enc_outputs, teacher_enc_hid = self.teacher.get_encoder()(
```

Comment:
I think in the old case of distilling to a copied student, this change uses more memory.
Trying to find a clean solution locally.

Reply:
I see. Not sure what part of the change makes it use more memory? Is there any way I can help?

Reply:
We used to only extract hidden states, now we extract hidden states and encoder outputs.
Your application doesn't need hidden states, the other application doesn't need encoder outputs, so splitting the call to `self.teacher.get_encoder()(` to only get what's needed is the next step.

Reply:
got it.I am attempting a fix.

Reply:
It is not true that the other application doesn't need encoder outputs. When calling the teacher decoder, if student and teacher encoders are different, the right thing to do is to pass the teacher encoder outputs, and as we can see from these lines, even in the original case, we could have `self.different_encoder` be true if not all layers were copied from teacher to student: https://github.com/huggingface/transformers/blob/02bdfc025169a301a57f48a7642c69de3d10a2db/examples/seq2seq/distillation.py#L60-L61.
So I think we do need to extract teacher encoder outputs inside this `if different_encoder` block.
However, we can extract hidden states optionally based on whether we need to calculate hidden loss or not. I have the following modified code which passes the unit tests. Please lmk what you think.
```
        if self.different_encoder:  # compute encoder hidden state loss
            with torch.no_grad():
                teacher_encoder = self.teacher.get_encoder()(
                    input_ids,
                    attention_mask=src_mask,
                    output_hidden_states=self.do_calc_hidden_loss,
                    return_dict=True
                )
            teacher_enc_outputs = teacher_encoder.last_hidden_state
            if self.do_calc_hidden_loss:
                teacher_enc_hid = teacher_encoder.hidden_states
                hid_loss_enc = self.calc_hidden_loss(
                    src_mask,
                    enc_hidden_state,
                    teacher_enc_hid,
                    self.e_matches,
                    normalize_hidden=self.hparams.normalize_hidden,
                )
```
That said, the same boolean `self.do_calc_hidden_loss` can be used to set `output_hidden_states` to `True` or `False` [in the call to the teacher decoder](https://github.com/huggingface/transformers/blob/02bdfc025169a301a57f48a7642c69de3d10a2db/examples/seq2seq/distillation.py#L176) as well . If the above makes sense to you I'll make a similar change here as well and push a commit.

Reply:
> It is not true that the other application doesn't need encoder outputs. 

The rationale was that if we are training the decoder to behave the same way, we should see what the 2 decoders produce given the same encoder outputs, even if the encoders are different.

Reply:
ah, thanks for the clarification! I'll take another look.

Reply:
Ok I have another fix that's passing unit tests. Please lmk what you think. I renamed the `self.do_calc_hidden_loss` to `self.different_base_models` because really that's what it was keeping track of, and the boolean value is flipped from the earlier one. 

```
        if self.different_encoder:  # compute encoder hidden state loss
            with torch.no_grad():
                teacher_encoder = self.teacher.get_encoder()(
                    input_ids,
                    attention_mask=src_mask,
                    output_hidden_states=not self.different_base_models,
                    return_dict=True
                )
            if self.different_base_models:
                teacher_enc_outputs = teacher_encoder.last_hidden_state
            else:
                teacher_enc_hid = teacher_encoder.hidden_states
                hid_loss_enc = self.calc_hidden_loss(
                    src_mask,
                    enc_hidden_state,
                    teacher_enc_hid,
                    self.e_matches,
                    normalize_hidden=self.hparams.normalize_hidden,
                )

        teacher_mask = input_ids.ne(pad_token_id)
        with torch.no_grad():
            outputs = self.teacher(
                input_ids,
                attention_mask=teacher_mask,
                encoder_outputs=(teacher_enc_outputs, ),
                decoder_input_ids=decoder_input_ids,
                lm_labels=labels,
                output_hidden_states=not self.different_base_models,
                return_dict=True,
            )
        dec_mask = decoder_input_ids.ne(pad_token_id)
        loss_ce = self.calc_ce_loss(dec_mask, lm_logits, outputs.logits)
        if (not self.different_base_models) and self.alpha_hid > 0:  # Intermediate supervision of decoder hidden states
            tdec_hidden = outputs.decoder_hidden_states
            hid_loss_dec = self.calc_hidden_loss(
                dec_mask, dec_hidden, tdec_hidden, self.d_matches, normalize_hidden=self.hparams.normalize_hidden
            )
```

`self.different_base_models` is initialized in `__init__` in the same place as the earlier `self.do_calc_hidden_loss` as follows:

```
        self.different_base_models = True

        if hparams.student is None or hparams.teacher == hparams.student:
            self.different_base_models = False
            # Intermediate supervision: Decide which layers to supervise
            if hparams.supervise_forward:
                self.e_matches = get_layers_to_supervise(n_student=len(self.e_layer_ids), n_teacher=teacher_encoder_layers)
                self.d_matches = get_layers_to_supervise(n_student=len(self.d_layer_ids), n_teacher=teacher_decoder_layers)
            else:  # student layer should emulate hidden states of the teacher layer it was copied from
                self.e_matches = self.e_layer_ids
                self.d_matches = self.d_layer_ids
```

Reply:
yes push that, I'll make some changes and get it merged.

Reply:
Thanks for iterating :)

Reply:
pushed!